diff -Nur xfce4-panel-profiles_old/Makefile.in.in xfce4-panel-profiles/Makefile.in.in
--- xfce4-panel-profiles_old/Makefile.in.in	2020-01-15 10:06:15.000000000 +0700
+++ xfce4-panel-profiles/Makefile.in.in	2020-11-12 16:36:08.440457451 +0700
@@ -12,6 +12,10 @@
 	chmod +x xfce4-panel-profiles.desktop
 	sed -e s,%prefix%,$(PREFIX), bin/$(APPNAME).in.in > bin/$(APPNAME).in
 	sed -e s,%python%,$(PYTHON), bin/$(APPNAME).in > bin/$(APPNAME)
+
+	sed -e s,%appname%,$(APPNAME), xfce4-panel-profiles/info.py.in.in > xfce4-panel-profiles/info.py.in
+	sed -e s,%version%,$(VERSION), xfce4-panel-profiles/info.py.in > xfce4-panel-profiles/info.py
+
 	chmod +x bin/$(APPNAME)
 
 locale/%/LC_MESSAGES/$(APPNAME).mo: po/%.po
@@ -39,6 +43,7 @@
 	install bin/$(APPNAME) $(DESTDIR)/$(PREFIX)/bin
 
 	install -d $(DESTDIR)/$(PREFIX)/share/$(APPNAME)/xfce4-panel-profiles
+	install xfce4-panel-profiles/info.py $(DESTDIR)/$(PREFIX)/share/$(APPNAME)/xfce4-panel-profiles
 	install xfce4-panel-profiles/panelconfig.py $(DESTDIR)/$(PREFIX)/share/$(APPNAME)/xfce4-panel-profiles
 	install xfce4-panel-profiles/xfce4-panel-profiles.py $(DESTDIR)/$(PREFIX)/share/$(APPNAME)/xfce4-panel-profiles
 	install xfce4-panel-profiles/xfce4-panel-profiles.glade $(DESTDIR)/$(PREFIX)/share/$(APPNAME)/xfce4-panel-profiles
@@ -94,3 +99,5 @@
 	rm -f Makefile.in
 	rm -f Makefile
 	rm -f xfce4-panel-profiles.1.gz
+	rm -f xfce4-panel-profiles/info.py
+	rm -f xfce4-panel-profiles/info.py.in
diff -Nur xfce4-panel-profiles_old/xfce4-panel-profiles/info.py.in.in xfce4-panel-profiles/xfce4-panel-profiles/info.py.in.in
--- xfce4-panel-profiles_old/xfce4-panel-profiles/info.py.in.in	1970-01-01 07:00:00.000000000 +0700
+++ xfce4-panel-profiles/xfce4-panel-profiles/info.py.in.in	2020-08-23 17:15:46.000000000 +0700
@@ -0,0 +1,2 @@
+appname = '%appname%'
+version = '%version%'
diff -Nur xfce4-panel-profiles_old/xfce4-panel-profiles/xfce4-panel-profiles.glade xfce4-panel-profiles/xfce4-panel-profiles/xfce4-panel-profiles.glade
--- xfce4-panel-profiles_old/xfce4-panel-profiles/xfce4-panel-profiles.glade	2020-01-15 10:06:15.000000000 +0700
+++ xfce4-panel-profiles/xfce4-panel-profiles/xfce4-panel-profiles.glade	2020-08-23 17:15:46.000000000 +0700
@@ -1,7 +1,8 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!-- Generated with glade 3.18.3 -->
+<!-- Generated with glade 3.22.2 -->
 <interface>
   <requires lib="gtk+" version="3.12"/>
+  <requires lib="libxfce4ui-2" version="4.12"/>
   <object class="GtkListStore" id="liststore1">
     <columns>
       <!-- column-name filename -->
@@ -12,97 +13,27 @@
       <column type="gint"/>
     </columns>
   </object>
-  <object class="GtkWindow" id="xfpanel_switch_window">
+  <object class="XfceTitledDialog" id="xfpanel_switch_window">
     <property name="can_focus">False</property>
+    <property name="title" translatable="yes">Panel Profiles</property>
     <property name="default_width">350</property>
     <property name="default_height">400</property>
     <property name="icon_name">xfce4-panel</property>
+    <property name="type_hint">dialog</property>
+    <property name="subtitle" translatable="yes">Backup and restore your panel configuration</property>
     <signal name="delete-event" handler="on_window_destroy" swapped="no"/>
-    <child>
+    <child internal-child="vbox">
       <object class="GtkBox" id="box1">
         <property name="visible">True</property>
         <property name="can_focus">False</property>
         <property name="orientation">vertical</property>
-        <child>
-          <object class="GtkBox" id="box3">
-            <property name="visible">True</property>
+        <child internal-child="action_area">
+          <object class="GtkButtonBox">
             <property name="can_focus">False</property>
-            <property name="hexpand">True</property>
-            <property name="vexpand">False</property>
-            <property name="orientation">vertical</property>
-            <child>
-              <object class="GtkEventBox" id="xfce_header">
-                <property name="name">xfce_header</property>
-                <property name="height_request">60</property>
-                <property name="visible">True</property>
-                <property name="can_focus">False</property>
-                <child>
-                  <object class="GtkBox" id="box2">
-                    <property name="visible">True</property>
-                    <property name="can_focus">False</property>
-                    <property name="hexpand">True</property>
-                    <property name="vexpand">True</property>
-                    <child>
-                      <object class="GtkImage" id="appicon">
-                        <property name="visible">True</property>
-                        <property name="can_focus">False</property>
-                        <property name="halign">start</property>
-                        <property name="margin_start">6</property>
-                        <property name="margin_top">1</property>
-                        <property name="pixel_size">48</property>
-                        <property name="icon_name">xfce4-panel</property>
-                        <property name="icon_size">6</property>
-                      </object>
-                      <packing>
-                        <property name="expand">False</property>
-                        <property name="fill">True</property>
-                        <property name="position">0</property>
-                      </packing>
-                    </child>
-                    <child>
-                      <object class="GtkLabel" id="apptitle">
-                        <property name="visible">True</property>
-                        <property name="can_focus">False</property>
-                        <property name="margin_start">12</property>
-                        <property name="label" translatable="yes">&lt;b&gt;&lt;big&gt;Panel Profiles&lt;/big&gt;&lt;/b&gt;
-Backup and restore your panel configuration</property>
-                        <property name="use_markup">True</property>
-                        <property name="ellipsize">end</property>
-                        <property name="xalign">0</property>
-                      </object>
-                      <packing>
-                        <property name="expand">True</property>
-                        <property name="fill">True</property>
-                        <property name="position">1</property>
-                      </packing>
-                    </child>
-                  </object>
-                </child>
-                <style>
-                  <class name="XfceHeading"/>
-                </style>
-              </object>
-              <packing>
-                <property name="expand">False</property>
-                <property name="fill">True</property>
-                <property name="position">0</property>
-              </packing>
-            </child>
-            <child>
-              <object class="GtkSeparator" id="separator1">
-                <property name="visible">True</property>
-                <property name="can_focus">False</property>
-              </object>
-              <packing>
-                <property name="expand">False</property>
-                <property name="fill">True</property>
-                <property name="position">1</property>
-              </packing>
-            </child>
           </object>
           <packing>
             <property name="expand">False</property>
-            <property name="fill">True</property>
+            <property name="fill">False</property>
             <property name="position">0</property>
           </packing>
         </child>
@@ -315,6 +246,9 @@
     <property name="icon_name">document-save-as</property>
     <property name="type_hint">dialog</property>
     <property name="transient_for">xfpanel_switch_window</property>
+    <child type="titlebar">
+      <placeholder/>
+    </child>
     <child internal-child="vbox">
       <object class="GtkBox" id="dialog-vbox1">
         <property name="can_focus">False</property>
diff -Nur xfce4-panel-profiles_old/xfce4-panel-profiles/xfce4-panel-profiles.py xfce4-panel-profiles/xfce4-panel-profiles/xfce4-panel-profiles.py
--- xfce4-panel-profiles_old/xfce4-panel-profiles/xfce4-panel-profiles.py	2020-01-15 10:06:15.000000000 +0700
+++ xfce4-panel-profiles/xfce4-panel-profiles/xfce4-panel-profiles.py	2020-08-23 17:15:46.000000000 +0700
@@ -32,11 +32,14 @@
 
 import gi
 gi.require_version('Gtk', '3.0')
+gi.require_version('libxfce4ui', '2.0')
 
-from gi.repository import Gtk, GLib, Gio
+from gi.repository import Gtk, GLib, Gio, libxfce4ui
 
 from panelconfig import PanelConfig
 
+import info
+
 warnings.filterwarnings("ignore")
 
 homedir = GLib.get_home_dir()
@@ -50,6 +53,9 @@
 
     def __init__(self):
         '''Initialize the Panel Profiles application.'''
+        # Temporary fix: https://stackoverflow.com/a/44230815
+        _ = libxfce4ui.TitledDialog()
+
         self.builder = Gtk.Builder()
         self.builder.set_translation_domain('xfce4-panel-profiles')
 
@@ -59,8 +65,6 @@
         self.builder.connect_signals(self)
 
         self.window = self.builder.get_object("xfpanel_switch_window")
-        self.window.set_title(_("Panel Profiles"))
-        self.fix_xfce_header()
 
         self.load_xfconf()
 
@@ -123,17 +127,6 @@
             interface,
             cancellable)
 
-    def fix_xfce_header(self):
-        ''' Set background-color of frame to base-color to make it resemble the
-        XfceHeading widget '''
-        self.xfce_header = self.builder.get_object("xfce_header")
-        entry = Gtk.Entry.new()
-        style = entry.get_style_context()
-        base_color = style.lookup_color("theme_base_color")
-        self.xfce_header.override_background_color(0, base_color[1])
-        fg_color = style.lookup_color("theme_fg_color")
-        self.xfce_header.override_color(0, fg_color[1])
-
     def get_data_dirs(self):
         dirs = []
         for directory in GLib.get_system_data_dirs():
@@ -309,10 +302,11 @@
         Gtk.main_quit()
 
     def on_help_clicked(self, *args):
-        '''Navigate to the application documentation on Help click.'''
-        screen = self.window.get_screen()
-        link = "https://docs.xfce.org/apps/xfce4-panel-profiles/start"
-        Gtk.show_uri(screen, link, Gtk.get_current_event_time())
+        '''Shows Xfce's standard help dialog.'''
+        libxfce4ui.dialog_show_help(parent=self.window,
+                                    component='xfce4-panel-profiles',
+                                    page='xfce4-panel-profiles',
+                                    offset=None)
 
 
 class PanelSaveDialog(Gtk.MessageDialog):
@@ -414,11 +408,14 @@
                 print(repr(e))
                 exit(1)
             exit(0)
+        elif sys.argv[1] == '--version':
+            print(info.appname + ' ' + info.version)
+            exit(0)
         else:
             print('Xfce Panel Profiles - Usage:')
-            print(sys.argv[0] + ' : load graphical user interface.')
-            print(sys.argv[0] + ' save <filename> : save current configuration.')
-            print(sys.argv[0] + ' load <filename> : load configuration from file.')
+            print(info.appname + ' : load graphical user interface.')
+            print(info.appname + ' save <filename> : save current configuration.')
+            print(info.appname + ' load <filename> : load configuration from file.')
             print('')
             exit(-1)
 
