diff -Nur libxfce4util-4.15.4_old/configure libxfce4util-4.15.4/configure
--- libxfce4util-4.15.4_old/configure	2020-11-03 07:42:34.000000000 +0700
+++ libxfce4util-4.15.4/configure	2020-11-20 10:48:08.812646059 +0700
@@ -3168,7 +3168,7 @@
 
 
 
-$as_echo "#define XFCE_VERSION_STRING \"4.16pre1\"" >>confdefs.h
+$as_echo "#define XFCE_VERSION_STRING \"4.16pre2\"" >>confdefs.h
 
 
 LIBXFCE4UTIL_VERSION_MAJOR=4
diff -Nur libxfce4util-4.15.4_old/configure.ac libxfce4util-4.15.4/configure.ac
--- libxfce4util-4.15.4_old/configure.ac	2020-11-03 07:42:30.000000000 +0700
+++ libxfce4util-4.15.4/configure.ac	2020-11-20 10:47:53.394647055 +0700
@@ -27,7 +27,7 @@
 dnl **************************************************************
 dnl *** Overall Xfce version, modify this when making releases ***
 dnl **************************************************************
-m4_define([xfce_version_string], [4.16pre1])
+m4_define([xfce_version_string], [4.16pre2])
 
 
 dnl # DO NOT MODIFY ANYTHING BELOW THIS LINE, UNLESS YOU KNOW WHAT
