From a07aeb66d462a1524735b75f6d219a20735f1ac5 Mon Sep 17 00:00:00 2001
From: Theo Linkspfeifer <lastonestanding@tutanota.com>
Date: Sun, 29 Nov 2020 22:31:00 +0100
Subject: mime: Accept Thunar's old id name (Fixes #247)

---
 dialogs/mime-settings/xfce-mime-helper.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dialogs/mime-settings/xfce-mime-helper.c b/dialogs/mime-settings/xfce-mime-helper.c
index b46e8517..47aeeb43 100644
--- a/dialogs/mime-settings/xfce-mime-helper.c
+++ b/dialogs/mime-settings/xfce-mime-helper.c
@@ -678,6 +678,11 @@ xfce_mime_helper_database_get_default (XfceMimeHelperDatabase *database,
       id = xfce_rc_read_entry_untranslated (rc, key, NULL);
       if (G_LIKELY (id != NULL))
         helper = xfce_mime_helper_database_lookup (database, category, id);
+
+      /* handle migrating from Xfce 4.14 or older */
+      if (helper == NULL && g_strcmp0 (id, "Thunar") == 0)
+        helper = xfce_mime_helper_database_lookup (database, category, "thunar");
+
       xfce_rc_close (rc);
       g_free (key);
     }
-- 
cgit v1.2.3

