From 856d73b51f5d237d3cf6792f5e9f55220e5facf1 Mon Sep 17 00:00:00 2001
From: Andre Miranda <andreldm@xfce.org>
Date: Sat, 19 Oct 2019 19:46:34 -0300
Subject: Fix build with panel 4.15

---
 src/smartbookmark.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/smartbookmark.c b/src/smartbookmark.c
index 8a89c24..554da24 100644
--- a/src/smartbookmark.c
+++ b/src/smartbookmark.c
@@ -41,7 +41,7 @@
 
 #include <libxfce4util/libxfce4util.h>
 #include <libxfce4ui/libxfce4ui.h>
-#include <libxfce4panel/xfce-panel-plugin.h>
+#include <libxfce4panel/libxfce4panel.h>
 
 /*
  * Types
-- 
cgit v1.2.3

