From ee3c336dd021e2065f1304c2885747c85f8dd50e Mon Sep 17 00:00:00 2001
From: Andre Miranda <andreldm@xfce.org>
Date: Fri, 4 Oct 2019 00:39:23 -0300
Subject: Allow compilation with panel 4.15

---
 panel-plugin/calculator.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/panel-plugin/calculator.c b/panel-plugin/calculator.c
index 0a8f30e..86c8bb5 100644
--- a/panel-plugin/calculator.c
+++ b/panel-plugin/calculator.c
@@ -28,7 +28,7 @@
 #include <gtk/gtk.h>
 #include <libxfce4util/libxfce4util.h>
 #include <libxfce4ui/libxfce4ui.h>
-#include <libxfce4panel/xfce-panel-plugin.h>
+#include <libxfce4panel/libxfce4panel.h>
 #include "parsetree.h"
 #include "parser.h"
 #include "eval.h"
@@ -211,7 +211,7 @@ static CalcPlugin *calc_new(XfcePanelPlugin *plugin)
     GtkOrientation orientation;
     GtkWidget *icon, *combo, *entry;
 
-    calc = panel_slice_new0(CalcPlugin);
+    calc = g_slice_new0(CalcPlugin);
     calc->plugin = plugin;
     calc_read_config(calc);
 
@@ -279,7 +279,7 @@ static void calc_free(XfcePanelPlugin *plugin, CalcPlugin *calc)
      * freeing of expr_hist enough?
      */
 
-    panel_slice_free(CalcPlugin, calc);
+    g_slice_free(CalcPlugin, calc);
 }
 
 
-- 
cgit v1.2.3

